messaging/src/main/java/org/axonframework/commandhandling/gateway/DefaultReactiveCommandGateway.java

```diff
@@ -0,0 +1,207 @@
+import org.axonframework.common.Registration;
+import org.axonframework.messaging.ReactiveMessageDispatchInterceptor;
+import reactor.core.publisher.Mono;
+
+import java.util.List;
+import java.util.concurrent.CopyOnWriteArrayList;
+
+import static java.util.Arrays.asList;
+import static org.axonframework.common.BuilderUtils.assertNonNull;
+
+/**
+ * Default implementation of {@link ReactiveCommandGateway}.
+ *
+ * @author Milan Savic
+ * @since 4.4
+ */
+public class DefaultReactiveCommandGateway implements ReactiveCommandGateway {
+
+    private final CommandBus commandBus;
+    private final RetryScheduler retryScheduler;
+    private final List<ReactiveMessageDispatchInterceptor<CommandMessage<?>>> dispatchInterceptors;
+
+    /**
+     * Creates an instance of {@link DefaultReactiveCommandGateway} based on the fields contained in the {@link
+     * Builder}.
+     * <p>
+     * Will assert that the {@link CommandBus} is not {@code null} and throws an {@link AxonConfigurationException} if
+     * it is.
+     * </p>
+     *
+     * @param builder the {@link Builder} used to instantiated a {@link DefaultReactiveCommandGateway} instance
+     */
+    protected DefaultReactiveCommandGateway(Builder builder) {
+        builder.validate();
+        this.commandBus = builder.commandBus;
+        this.retryScheduler = builder.retryScheduler;
+        this.dispatchInterceptors = builder.dispatchInterceptors;
+    }
+
+    /**
+     * Instantiate a Builder to be able to create a {@link DefaultReactiveCommandGateway}.
+     * <p>
+     * The {@code dispatchInterceptors} are defaulted to an empty list.
+     * The {@link CommandBus} is a <b>hard requirements</b> and as such should be provided.
+     * </p>
+     *
+     * @return a Builder to be able to create a {@link DefaultReactiveCommandGateway}
+     */
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    @Override
+    public <R> Mono<R> send(Mono<Object> command) {
+        return processInterceptors(command.map(GenericCommandMessage::asCommandMessage))
+                .flatMap(commandMessage -> Mono.create(
+                        sink -> {
+                            CommandCallback<Object, R> commandCallback = (cm, result) -> {
+                                try {
+                                    if (result.isExceptional()) {
+                                        sink.error(result.exceptionResult());
+                                    } else {
+                                        sink.success(result.getPayload());
+                                    }
+                                } catch (Exception e) {
+                                    sink.error(e);
+                                }
+                            };
+                            if (retryScheduler != null) {
+                                commandCallback = new RetryingCallback<>(commandCallback, retryScheduler, commandBus);
```

Comment:
Is there a use case where a `retryScheduler` is needed and Mono's out of the box retry mechanism isn't enough? If so, documentation for each use case would be great.

Reply:
Mono retry mechanism is great. Retry scheduler, in this case, is an optimization (we might discuss whether is necessary) to retry only command sending and not the whole loop with interceptors. Also, you might wanna reconsider Mono's retry since you'd retry interceptors which purpose is to return `Mono.error` which indicates command dispatching should be cancelled.